### ISSUE TITLE
Short data type badly needed

### ISSUE THREAD

**Issue by cmcdee** [ceylon/ceylon.language#795]
I've moved Java app that analyze a large amount of telemetry data to Ceylon. The largest dataset has datapoints in the [0..20000] range, a few million rows at most. Datapoints would fit neatly in a short data type, even a signed one. The data is squeezed into a collection before analysis, because multiple passes are needed, and doing it diskbased is just too slow.

With Ceylons 64-bit Integers, an immense amount of memory is vasted, and we eventually run out of it (hard to deal with since https://github.com/ceylon/ceylon-runtime/issues/55 still isn't fixed, as far as I can tell)

**Comment by jvasileff** [comment]
@cmcdee it is possible to create and use a Java `short[]`, see https://modules.ceylon-lang.org/repo/1/ceylon/interop/java/1.2.0/module-doc/api/index.html#createJavaShortArray

You will of course incur some primitive conversions between `long` and `short` on usage, but it sounds like your concern is primarily storage.

**Comment by cmcdee** [comment]
Performance matters.

**Comment by jvasileff** [comment]
How much? It may be worth at try; casting between primitive types on the JVM should be very fast.

**Comment by cmcdee** [comment]
Might be. That aside, not beeing able to express the value domain with the correct datatype is a language weakness. Why should I need to drop into Java for such a simple thing.

**Comment by quintesse** [comment]
You see it as a weakness, we see it as an advantage. The language is much more streamlined without all those extra types that are only useful under a limited set of circumstances and don't even make sense on certain platforms. And the whole idea of having Java interop is that you are still able to do the all the things that Java can if you really need them.

**Comment by jvasileff** [comment]
How often do you really need two's complement overflow behavior on a 16 bit integer type?

**Comment by cmcdee** [comment]
@quintesse You answered my real question. Ceylon is a language that depends on Java to do certain basic things. Fair enough. It's a decision (a really bad one)

**Comment by quintesse** [comment]
@cmcdee you might have used `BigDecimal` at some point, or `BitSet`, these are both perfectly reasonable abstractions to work around the fact that Java doesn't have a native bit type of arbitrary length nor does it have a number type of arbitrary precision. In Ceylon there's only 2 numeric types, but that doesn't mean you couldn't write an abstraction around Java shorts. Btw, there *is* a `ShortArray` available in Ceylon.

**Comment by cmcdee** [comment]
Look. When dealing with, say, protocol parsing, the natural way to deal with a 16-bit numeric entity is to use a 16 bit numeric data type. The natural way to deal with 32-bit numeric entities is using 32-bit numeric data types. Almost all statically typed languages support this, and people exploit them to make things efficient in terms of both speed and memory.

Yes, you can make it work in Ceylon, but you have to resort to tricks, and that's just sad. The language wouldn't be less streamlined with the addition of these commonly used data types (if that's the case, then something is wrong with the type system... but I'm sure the narrowing, sign-extended widening can be made to work just fine)

**Comment by ASzc** [comment]
Note that `Integer` isn't 64 bits on all backends, on JS `integerAddressableSize` is 32 bits. Size is a platform-specific detail, so it's quite natural that you would need `native` code to handle specific sizes. It would be disingenuous for Ceylon to offer Short or Long on JS.

I know [from experience](https://github.com/ASzc/svalinn/tree/master/source/ca/szc/svalinn/compressor) that low-level stuff like this is difficult. I agree more effort should be made in that area, but it shouldn't compromise the primary aims of Ceylon.

**Comment by gavinking** [comment]
Alright, so everyone calm down here :-)

Look, it has _always_ been the intent to allow optimized platform-specific numeric types via use of an annotation. Since the very earliest revisions of the language specification the spec has mentioned a `small` annotation, which may be used to specify the bit width of a numeric field, for example:

```ceylon
small Integer int;
```

Would be a 32-bit Integer field. Why would we make this an annotation instead of a type, which is more traditional? Well, because:

- I don't see any value in encoding this information into the type system. I don't see how the type system can do any useful reasoning around numeric overflow.
- In a cross-platform language, it's always by nature going to be a _hint_.
- Low-level graphics or sound programming is not a typical use for the JVM, JavaScript VMs, or for the Dart VM.
- Finally, on the JVM, with a modern 64-bit processor, it's not even very clear that there is much difference at all between `short`, `int`, and `long`, until you start creating huge primitive arrays full of them. So these types only really become interesting when you start looking at _container types_.

However, I now have some reasons to wonder whether this original reasoning might not be quite right:

- Annotation information is not propagated, which might in practice result in too much verbosity for a program which does a *lot* of work with, say, 16-bit integers.
- If @sadmac7000 finally gives us a native compiler for Ceylon, there will be lot more interesting applications for these types.
- Annotations don't help when working with generic container types like `Array`, where today you would need to use an `Array<java.lang::Short>` to stick stuff in a primitive array of type `short[]`. In practice what this means is that you would actually use Java's `ShortArray`.

I'm therefore open to the idea that for certain kinds of low-level code, we should provide a package of platform-dependent numeric types that are optimized by the compiler for a certain platform, along with facilities for creating arrays of each. Such a package wouldn't be for everyday business programming in Ceylon, it wouldn't be part of the language module, and it would not be cross-platform.

However, the right thing to do here is going to be very affected by what happens in Java 9 with the work on generics over primitive types. It's at least *possible* that the problem of abstracting container types over numeric types will be solved almost for free by the Java 9 VM. (And our future native compiler will hopefully be able to also solve this.)

So I don't believe that this is the right moment to bite off this problem. 

Which, I think, is perfectly fine, since:

- For now we already have `ShortArray`, `IntArray`, `FloatArray` and friends which offer efficient containers on the JVM.
- I agree with others here that there is no problem with `short`s being converted to `long`s when they are read out of the `ShortArray`. Surely you're running on a 64-bit architecture anyway! 
- And I actively disbelieve that anyone is actually using `short`s to get 16-bit unsigned&mdash;i.e. mod-65535&mdash;modular arithmetic!

In conclusion, I think the need for 16-bit and 32-bit numeric types is actually really a need for _containers_ of these types, and that, given the limitations of the Java 8 VM, that this need cannot be addressed elegantly at this time. And that the current _inelegant_ solution (`ShortArray`, `IntArray`, `FloatArray`) is good enough.

If @sadmac7000 delivers us a working compiler for LLVM, or if Valhalla finally ends up giving us the abstractions we need, that whole calculation changes.

**Comment by gavinking** [comment]
> @jvasileff I often need serialization of and arithmetic on the smallest datatype required for a given dataset. So far only Javascript and Ceylon has failed to give me that.

@cmcdee please explain this further, since, like @jvasileff I find it difficult to believe that mod-65535 arithmetic is a property of `short`s that is useful to you. It seems to me that the arithmetic properties of `short` must be inconsequential to your needs, and that what your real needs are related to storage.

**Comment by cmcdee** [comment]
> I don't see any value in encoding this information into the type system. I don't see how the type system can do any useful reasoning around numeric overflow

This is important information to certain compiler optimizations, forwarding this to the llvm type system will be crucial. The Java 8 JIT does the same things on the short type when it can.

> @cmcdee please explain this further, since, like @jvasileff, I find it difficult to believe that mod-65535 arithmetic is a property of shorts that is useful to you. It seems to me that the arithmetic properties of short must be inconsequential to your needs, and that what your real needs are related to storage.

Arithmetic on short is just as useful as arithmetic on 64-bit integers. There are value domains where short is the right data type; one example is a array of coordinates you want to manipulate to do image transforms. short is always the right type. Byte it too small, 32-bit ints too large. I don't understand why this is hard to understand.

Also, Java will have packed objects (which will be even better than http://objectlayout.org/) where you can express the same as __packed__ struct { short protocol, short port, int version, char term } When mapping an incoming packet to a Ceylon class, you want the right data type, the right alignment and the right field ordering.

Obviously you want to do arithmetics and numeric comparisons on some of these fields efficiently, without boxing.

**Comment by gavinking** [comment]
> This is important information to certain compiler optimizations, 

Examples?

> forwarding this to the llvm type system will be crucial. The Java 8 JIT does the same things on the short type when it can.

Well that would still be possible with the annotation-based approach. Whether the information is encoded into the type system, or via some other mechanism is not critical here.

> Arithmetic on short is just as useful as arithmetic on 64-bit integers. There are value domains where short is the right data type; 

You've misunderstood the point. What we're saying is not that if I have a `short`, I don't want to do arithmetic on it. What we're saying is that _there are no special properties that `short` has that `long` doesn't also have_.

> one example is an array of coordinates you want to manipulate to do image transforms. short is always the right type. Byte it too small, 32-bit ints too large. I don't understand why this is hard to understand.

I don't see this at all. _Why_ is a 32-bit int too large? To be clear: this is a 64-bit CPU. AFAIK, there are no 16-bit registers. This `short` thingy is a language-level fiction.

**Comment by gavinking** [comment]
> Also, Java will have packed objects 

Well now you're just speculating about functionality that Java might have in the future. That's not a design constraint on Ceylon *today*.

**Comment by cmcdee** [comment]
> Well now you're just speculating about functionality that Java might have in the future. That's not a design constraint on Ceylon today.

Sigh, just trying to make the case that languages see the need for these things, so you might as well deal with it now.

> I don't see this at all. Why is a 32-bit int too large? To be clear: this is a 64-bit CPU

You don't understand that a compiler backend can use type information to perform optimizations? You don't appreciate the importance of fitting data in a cacheline? I guess Ceylon is not that kind of language. Basically just another Python or Ruby. Good to know.

**Comment by gavinking** [comment]
> You don't understand that a compiler backend can use type information to perform optimizations? You don't appreciate the importance of fitting data in a cacheline? I

Well, sure I do. I see it for something like the LLVM backend for Ceylon. I _don't_ see it for the JVM or JavaScript VMs.

>  I guess Ceylon is not that kind of language. Basically just another Python or Ruby.

Ceylon is constrained by the nature of the platform it is running on. It's of course *much* faster than Python or Ruby when running on the JVM, but no, it doesn't support the sort of fine-grained optimization you might expect in say C or Rust.

As I said: *if and when* we have a working LLVM backend, then the whole nature of this discussion changes. We don't have it today.

**Comment by RossTate** [comment]
Having implemented a superoptimizing compiler, small size data tends to
decrease performance because the assembly needs to add instructions to
force precise recreation of small data that the hardware doesn't
automatically do. Padding to fit the size of the hardware representation
and ignoring differences is overflow and such tends to actually improve
performance. So size tends to only matter for storage in memory.

Note I say "tends to" a lot. There are some advanced optimizations only
possible for small data, but optimizers often can't figure them out. More
commonly, some devices will have hardware specialized to smaller data in
order to achieve better performance and lower energy costs. Nonetheless,
these are the less common cases.

**Comment by cmcdee** [comment]
@RossTate so "matters for storage in memory" and performance is closely correlated. "superoptimizing" compiler are very much trying to fit things in cachelines these days, which is why the platform triple is useful information so they know how long this line actually is. Data sizes matter very much in this case for performance reasons...

**Comment by gavinking** [comment]
> Nothing would stop you from lowering a Short sequence to a fast and compact jvm ShortArray.

Well we _already do this_ for an `Array<java.lang:Short>`! Try it out you will get a Java `short[]` under the covers.

But the problem is that when iterating this thing, or accessing elements of it, you will get *boxed* `java.lang.Short` instances. That's a problem that perhaps Valhalla will solve, or that *perhaps* we can solve in for at least some special at the compiler level. But neither of those solutions is going to happen right now.

**Comment by cmcdee** [comment]
> But neither of those solutions is going to happen right now.

Noted.

**Comment by RossTate** [comment]
"Data sizes matter very much in this case for performance reasons." Right.
So I think that we're all in agreement that size matters for integers
stored in memory, but doesn't matter much for intermediate variables.
Ceylon has been addressing this by not having small integer types, but by
having small integer data structures. More could be done, and no one's
closed to doing that. For example, Ceylon could enable programmers to
annotate fields with size information so that data structures can be stored
more compactly.
